src/blister.h

```diff
@@ -352,6 +355,147 @@ extern void ConvBlist (
 
 /****************************************************************************
 **
+*F  CopyBits( <fromblock>, <from-starting-bit>, <toblock>, <to-starting-bit>,
+**            <numbits> )
+**
+**  `CopyBits' copies <numbits> bits (numbering bits within a UInt
+**   from the least significant to the most significant) starting with
+**   bit number <from-starting-bit> of UInt *<fromblock> to a destination
+**   starting at bit <to-starting-bit> of *<toblock>.
+**
+**   The full function is provided as an inline function here because it is
+**   used in both blister.c and vecgf2.c and the compiler can do significant
+**   optimisations if some or all of <frombits>, <tobits> and <numbits> are
+**   known at compile time.
+*/
+
+/* constructs a mask that selects bits <from> to <to> inclusive of a UInt */
+static inline UInt mask(UInt from, UInt to)
+{
+    return ((to == BIPEB - 1) ? 0 : (1L << (to + 1))) - (1L << from);
+}
+
+/* copies a block of bits from the UInt <from> to the one pointed at
+   by <to> the block starts in <from> at <startbit> and ends at
+   <endbit>. It is shifted left by <shift> and must still fit within a
+   single word (so endbits + shift must be < BIPEB and frombits +
+   shift must be non-negative */
+
+static inline void
+CopyInWord(UInt * to, UInt startbit, UInt endbit, UInt from, Int shift)
+{
+    UInt m = mask(startbit + shift, endbit + shift);
+    *to &= ~m;
+    if (shift >= 0)
+        *to |= ((from << shift) & m);
+    else
+        *to |= ((from >> -shift) & m);
+}
+
+
+static inline  __attribute__((always_inline)) void CopyBits(const UInt * fromblock,
+              UInt         frombit,
+              UInt *       toblock,
+              UInt         tobit,
+              UInt         nbits)
+{
+    UInt tailbits;
+    UInt x;
+    UInt wholeblocks;
+    if (!nbits)
+        return;
+    /* Normalize */
+    fromblock += frombit / BIPEB;
+    frombit %= BIPEB;
+    toblock += tobit / BIPEB;
+    tobit %= BIPEB;
+    /* If the alignment of the two data blocks matches, things are relatively
+     * easy
+     */
+    if (frombit == tobit) {
+        /* if the first and last words are the same word */
+        if ((frombit + nbits) < BIPEB) {
+            CopyInWord(toblock, frombit, frombit + nbits - 1, *fromblock, 0);
+            return;
+        }
+        /* do we need to start by copying a partial word */
+        if (frombit) {
+            CopyInWord(toblock, frombit, BIPEB - 1, *fromblock, 0);
+            fromblock++;
+            toblock++;
+            nbits -= (BIPEB - frombit);
+            frombit = 0;
+            tobit = 0;
+        }
+        /* Now move whole words */
+        if ((wholeblocks = nbits / BIPEB))
+            memcpy((void *)toblock, (void *)fromblock,
```

Comment:
I don't think you need the typecasts here (and the second one removes the `const` qualifier of `fromblock`.